### PR TITLE
refact(upgrade): modified the upgrade code for 1.3.0 upgrades

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham <shubham.bajpai@mayadata.io>

**What this PR does / why we need it**:
This PR adds the ability to upgrade from 1.0.0, 1.1.0, 1.2.0 to 1.3.0. The upgrade steps wait for the new schema changes introduced in 1.3.0 and use those VersionDetails spec to upgrade the resources through the controller. The job sets the Desired version after the patching of corresponding deployment is done.

The PR is also includes changes that will allow upgrades from RC builds for different versions like `1.2.0-RC1 to 1.3.0-RC1` and upgrades from RC builds within the same version 
`1.3.0-RC1 to 1.3.0-RC2` and `1.3.0-RC1 to 1.3.0`.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go [hunk 1]
```diff
@@ -231,6 +235,19 @@ func (c *cstorCSPOptions) preUpgrade(cspName, openebsNamespace string) error {
...
+			return err
+		}
+		c.cspObj = obj
+		// Sleep equal to the default sync time
+		time.Sleep(30 * time.Second)
+	}
+	return nil
+}
+
 func (c *cstorCSPOptions) poolInstanceUpgarde(openebsNamespace string) error {
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: spelling mistake here

File: cmd/upgrade/executor/options.go [hunk 2]
```diff
@@ -75,3 +77,33 @@ func (u *UpgradeOptions) InitializeDefaults(cmd *cobra.Command) error {
 
 	return nil
 }
+
+// get the path for the version
```
Comment (vishnuitta) [inline cmd/upgrade/executor/options.go]: please add comments on what we are trying to do

File: cmd/upgrade/executor/options.go [hunk 3]
```diff
@@ -75,3 +77,33 @@ func (u *UpgradeOptions) InitializeDefaults(cmd *cobra.Command) error {
...
+		)
+	if err != nil {
+		return "", err
+	}
+	if len(mayaPods.Items) != 1 {
+		return "", errors.Errorf("Expecting 1 maya pod got %d", len(mayaPods.Items))
+	}
+	v := strings.Split(mayaPods.Items[0].Labels["openebs.io/version"], "-")[0]
+
+	if from == v && to == v {
```
Comment (vishnuitta) [inline cmd/upgrade/executor/options.go]: add comments here for the logic here

File: cmd/upgrade/executor/cstor_volume.go [hunk 4]
```diff
@@ -77,20 +76,13 @@ func (u *UpgradeOptions) RunCStorVolumeUpgradeChecks(cmd *cobra.Command) error {
...
-	case "1.0.0-1.1.0", "1.0.0-1.2.0", "1.1.0-1.2.0":
+	path, err := u.getUpgradePath()
+	if err != nil {
+		return err
+	}
+	switch path {
+	case "1.0.0-1.3.0", "1.1.0-1.3.0", "1.2.0-1.3.0":
+		// RC1-RC2 for RC1 to RC2, RC1- for RC1 to GA, RC2- for RC2 to GA
 		klog.Infof("Upgrading to %s", u.toVersion)
 		err := upgrade100to120.Exec(u.fromVersion, u.toVersion,
```
Comment (vishnuitta) [inline cmd/upgrade/executor/cstor_volume.go]: this need to be changed

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go [hunk 5]
```diff
@@ -113,7 +114,7 @@ func patchCSP(cspObj *apis.CStorPool) error {
 	cspVersion := cspObj.Labels["openebs.io/version"]
 	if cspVersion == currentVersion {
 		tmpl, err := template.New("cspPatch").
-			Parse(templates.OpenebsVersionPatch)
+			Parse(templates.VersionDetailsPatch)
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: `patch` is being done later in the code. Will that stay or will that be changed to `update`?

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go [hunk 6]
```diff
@@ -252,6 +266,17 @@ func (c *cstorCSPOptions) poolInstanceUpgarde(openebsNamespace string) error {
...
+		statusObj.Message = "failed to verify versiondetails for cstor pool"
+		statusObj.Reason = strings.Replace(err.Error(), ":", "", -1)
+		c.utaskObj, uerr = updateUpgradeDetailedStatus(c.utaskObj, statusObj, openebsNamespace)
+		if uerr != nil && isENVPresent {
+			return uerr
+		}
+		return err
+	}
+
 	err = patchCSP(c.cspObj)
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: don't we need to wait for `current` to become `desired`?

File: pkg/upgrade/templates/v1/common.go [hunk 7]
```diff
@@ -24,5 +24,16 @@ var (
...
+ }`
+	// VersionDetailsPatch is generic template for version details patch
+	VersionDetailsPatch = `{
+	"metadata": {
+	   "labels": {
+		  "openebs.io/version": "{{.}}"
+	   }
+	},
+	"versionDetails": {
+		"desired": "{{.}}"
```
Comment (vishnuitta) [inline pkg/upgrade/templates/v1/common.go]: is this supposed to be `Desired`?

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/spc_upgrade.go [hunk 8]
```diff
@@ -72,6 +73,24 @@ func spcUpgrade(spcName, openebsNamespace string) (*utask.UpgradeTask, error) {
...
 	return nil, nil
 }
+
+func updateSPC(name string) error {
+	client := spc.NewKubeClient()
+	spcObj, err := client.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	spcObj.VersionDetails.Desired = upgradeVersion
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/spc_upgrade.go]: why only Desired? We need to set for other fields as well?

File: cmd/upgrade/executor/options.go (hunk 2 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/options.go]: Done

File: cmd/upgrade/executor/options.go (hunk 3 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/options.go]: Done

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go (hunk 5 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: Patch will stay.

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go (hunk 6 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: added the wait function

File: pkg/upgrade/templates/v1/common.go (hunk 7 above)
Comment (shubham14bajpai) [reply pkg/upgrade/templates/v1/common.go]: in json it will be `desired` only

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/spc_upgrade.go (hunk 8 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/spc_upgrade.go]: only desired is set by the job all other fields are populated by the cstor-operator itself.

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go [hunk 9]
```diff
@@ -231,7 +232,33 @@ func (c *cstorCSPOptions) preUpgrade(cspName, openebsNamespace string) error {
...
-func (c *cstorCSPOptions) poolInstanceUpgarde(openebsNamespace string) error {
+func (c *cstorCSPOptions) waitForCSPCurrentVersion() error {
+	for c.cspObj.VersionDetails.Current == "" {
+		obj, err := cspClient.Get(c.cspObj.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		c.cspObj = obj
+		// Sleep equal to the default sync time
+		time.Sleep(30 * time.Second)
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: add a log here..
also, add some comments..

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go [hunk 10]
```diff
@@ -231,7 +232,33 @@ func (c *cstorCSPOptions) preUpgrade(cspName, openebsNamespace string) error {
...
+	}
+	return nil
+}
+
+func (c *cstorCSPOptions) verifyCSPVersionReconcile() error {
+	for c.cspObj.VersionDetails.Current != c.cspObj.VersionDetails.Desired {
+		obj, err := cspClient.Get(c.cspObj.Name, metav1.GetOptions{})
```
Comment (vishnuitta) [inline pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: move the sleep to the start of loop

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go (hunk 1 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: Corrected

File: cmd/upgrade/executor/cstor_volume.go (hunk 4 above)
Comment (shubham14bajpai) [reply cmd/upgrade/executor/cstor_volume.go]: Done

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go (hunk 9 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: Done

File: pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go (hunk 10 above)
Comment (shubham14bajpai) [reply pkg/upgrade/1.0.0-1.1.0/v1alpha1/csp_upgrade.go]: Done

**Pull request merged**